python/ray/data/datasource/file_based_datasource.py

```diff
@@ -256,10 +259,15 @@ def read_files(
                     read_path,
                     lambda: open_input_source(fs, read_path, **open_stream_args),
                 ) as f:
-                    for data in read_stream(f, read_path):
+                    for block in read_stream(f, read_path):
                         if partitions:
-                            data = _add_partitions(data, partitions)
-                        yield data
+                            block = _add_partitions(block, partitions)
+                        if self._include_paths:
+                            block_accessor = BlockAccessor.for_block(block)
+                            block = block_accessor.append_column(
```

Comment:
`append_column` performs a copy. We can avoid the copy by duplicating the logic in each subclasses' `_read_file` implementation:

https://github.com/ray-project/ray/blob/2a46f26f57cb7a0fb6afa633735c14fa517dc825/python/ray/data/datasource/binary_datasource.py#L31-L34

Alternatively, if `_read_stream` yields data batches and we mutate the batches, we can also avoid a copy.

Given that "premature optimization is the root of all evil", I've opted for the approach that minimizes code duplication. We can revisit this as part of https://github.com/ray-project/ray/issues/39078.